### PR TITLE
Shuffles gen args back to cmd and operations to prevent dep-graph issues

### PR DESCRIPTION
**Pull request by timothysc**
This commit shuffles the args back closer to the commands and moves around some of the objects. 

I isolated at just gen and added a bunch of //TODOs for follow-on PRs.  

Signed-off-by: Timothy St. Clair <timothysc@gmail.com>

### CODE REVIEW

File: cmd/sonobuoy/app/gen.go [hunk 1]
```diff
@@ -39,21 +42,40 @@ var GenCommand = &cobra.Command{
...
-	args.AddModeFlag(&genopts.ModeName, GenCommand)
-	args.AddNamespaceFlag(&genopts.Namespace, GenCommand)
-
+	AddGenFlags(&genopts, GenCommand)
 	RootCmd.AddCommand(GenCommand)
 }
 
+// AddGenFlags adds generation flags to a command
+func AddGenFlags(gen *ops.GenConfig, cmd *cobra.Command) {
+	cmd.PersistentFlags().StringVarP(
```
Comment (chuckha) [inline cmd/sonobuoy/app/gen.go]: Do these need to be persistent flags? That to me implies there is a subcommand and this is a global option flag for this command and all children commands.

File: cmd/sonobuoy/app/gen.go [hunk 2]
```diff
@@ -39,21 +42,40 @@ var GenCommand = &cobra.Command{
...
+	)
+	// TODO: This variable default needs saner image defaults from ops.f(n) or config
+	cmd.PersistentFlags().StringVarP(
+		&gen.Namespace, "namespace", "n", "heptio-sonobuoy",
+		"The namespace to run Sonobuoy in. Only one Sonobuoy run can exist per namespace simultaneously.",
+	)
+	// TODO: Need to provide ability to override config structure and allow for sane defaults
+	// TODO: Need to provide ability to override e2e-focus
+	// TODO: Need to provide ability to override e2e-skip
+}
```
Comment (chuckha) [inline cmd/sonobuoy/app/gen.go]: Out of scope for this PR, but what do you think about being able to include/exclude plugins with this command? something like `sonobuoy gen --plugins=e2e --e2e-mode=quick` (Then we get into a weird space where a plugin author could integrate with the sonobuoy command by providing a flag set to add their plugin flags `--plugins=e2e,bulkhead --bulkhead-option=abcd`).

File: cmd/sonobuoy/app/gen.go [hunk 3]
```diff
@@ -39,21 +42,40 @@ var GenCommand = &cobra.Command{
...
 	RootCmd.AddCommand(GenCommand)
 }
 
+// AddGenFlags adds generation flags to a command
+func AddGenFlags(gen *ops.GenConfig, cmd *cobra.Command) {
+	cmd.PersistentFlags().StringVarP(
+		&mode, "e2e-mode", "", string(ops.Conformance),
+		fmt.Sprintf("What mode to run sonobuoy in. [%s]", strings.Join(ops.GetModes(), ", ")),
+	)
+	// TODO: This variable default needs saner image defaults from ops.f(n) or config
```
Comment (chuckha) [inline cmd/sonobuoy/app/gen.go]: We've been trying to keep the suggestion provided by @davecheney to annotate TODOs with people so context can be gleaned for future developers. Adding a name does not at all imply ownership, simply the name of the person that wrote the comment. If we all like this idea we should formalize it along with other decisions we've made (this is on my backlog).

example:
`// TODO(chuckha) make this code more betterer.`

File: cmd/sonobuoy/app/gen.go (hunk 1 above)
Comment (timothysc) [reply cmd/sonobuoy/app/gen.go]: Doesn't hurt if we somehow expanded, but if they were local then we would need to shuffle them all.

File: cmd/sonobuoy/app/gen.go (hunk 2 above)
Comment (timothysc) [reply cmd/sonobuoy/app/gen.go]: Let's tackle the plugin question separately, I have thoughts there and we should leverage config.

File: cmd/sonobuoy/app/operations/gen.go [hunk 4]
```diff
@@ -45,10 +44,10 @@ type templateValues struct {
 }
 
 // GenerateManifest fills in a template with a Sonobuoy config
-func GenerateManifest(cfg GenConfig) ([]byte, error) {
-	mode := cfg.ModeName.Get()
+func (g *GenConfig) GenerateManifest() ([]byte, error) {
```
Comment (liztio) [inline cmd/sonobuoy/app/operations/gen.go]: I don't know that I like this. Having a `Config` object have methods to execute feels weird. What does this gain us over the old style?

File: cmd/sonobuoy/app/gen.go [hunk 5]
```diff
@@ -19,16 +19,19 @@ package app
...
 
-	"github.com/heptio/sonobuoy/cmd/sonobuoy/app/args"
 	ops "github.com/heptio/sonobuoy/cmd/sonobuoy/app/operations"
 	"github.com/heptio/sonobuoy/pkg/errlog"
 )
 
 var genopts ops.GenConfig
+var mode string
+
+type foo string
```
Comment (liztio) [inline cmd/sonobuoy/app/gen.go]: what's this for?

File: cmd/sonobuoy/app/gen.go [hunk 6]
```diff
@@ -39,21 +42,40 @@ var GenCommand = &cobra.Command{
...
+	)
+	// TODO: Need to provide ability to override config structure and allow for sane defaults
+	// TODO: Need to provide ability to override e2e-focus
+	// TODO: Need to provide ability to override e2e-skip
+}
+
 func genManifest(cmd *cobra.Command, args []string) {
-	code := 0
-	bytes, err := ops.GenerateManifest(genopts)
+	err := genopts.ModeName.Set(mode)
```
Comment (liztio) [inline cmd/sonobuoy/app/gen.go]: this could be collapsed into a single if statement.

File: cmd/sonobuoy/app/gen.go [hunk 7]
```diff
@@ -19,16 +19,19 @@ package app
...
 
 	"github.com/pkg/errors"
 	"github.com/spf13/cobra"
 
-	"github.com/heptio/sonobuoy/cmd/sonobuoy/app/args"
 	ops "github.com/heptio/sonobuoy/cmd/sonobuoy/app/operations"
 	"github.com/heptio/sonobuoy/pkg/errlog"
 )
 
 var genopts ops.GenConfig
```
Comment (liztio) [inline cmd/sonobuoy/app/gen.go]: I don't think this needs to be a global variable anymore. It only was before so we could use it from `init`.

File: cmd/sonobuoy/app/gen.go [hunk 8]
```diff
@@ -39,21 +42,40 @@ var GenCommand = &cobra.Command{
...
 }
 
+// AddGenFlags adds generation flags to a command
+func AddGenFlags(gen *ops.GenConfig, cmd *cobra.Command) {
+	cmd.PersistentFlags().StringVarP(
+		&mode, "e2e-mode", "", string(ops.Conformance),
+		fmt.Sprintf("What mode to run sonobuoy in. [%s]", strings.Join(ops.GetModes(), ", ")),
+	)
+	// TODO: This variable default needs saner image defaults from ops.f(n) or config
+	cmd.PersistentFlags().StringVarP(
```
Comment (liztio) [inline cmd/sonobuoy/app/gen.go]: I think it'd be better to either use gen.* everywhere or nowhere. Makes it easier to know where to add new options.

File: cmd/sonobuoy/app/gen.go [hunk 9]
```diff
@@ -39,21 +42,40 @@ var GenCommand = &cobra.Command{
...
+	// TODO: Need to provide ability to override config structure and allow for sane defaults
+	// TODO: Need to provide ability to override e2e-focus
+	// TODO: Need to provide ability to override e2e-skip
+}
+
 func genManifest(cmd *cobra.Command, args []string) {
-	code := 0
-	bytes, err := ops.GenerateManifest(genopts)
+	err := genopts.ModeName.Set(mode)
 	if err == nil {
```
Comment (liztio) [inline cmd/sonobuoy/app/gen.go]: this could be the more familiar `if err != nil { os.exit() }`

File: cmd/sonobuoy/app/run.go [hunk 10]
```diff
@@ -36,9 +36,9 @@ func init() {
 		Run:   submitSonobuoyRun,
 		Args:  cobra.ExactArgs(0),
 	}
-	args.AddModeFlag(&runopts.GenConfig.ModeName, cmd)
-	args.AddSonobuoyImageFlag(&runopts.GenConfig.Image, cmd)
-	args.AddNamespaceFlag(&runopts.GenConfig.Namespace, cmd)
+	AddGenFlags(&runopts.GenConfig, cmd)
+
+	// TODO: reduce this too.
```
Comment (liztio) [inline cmd/sonobuoy/app/run.go]: what do you mean by `reduce`?

File: cmd/sonobuoy/app/gen.go (hunk 3 above)
Comment (timothysc) [reply cmd/sonobuoy/app/gen.go]: Updated.

File: cmd/sonobuoy/app/operations/run.go [hunk 11]
```diff
@@ -20,7 +20,9 @@ import (
 	"bytes"
 	"io"
 
+	// TODO - remove kubeconfig too.
```
Comment (liztio) [inline cmd/sonobuoy/app/operations/run.go]: We need a kubeconfig somewhere so we can actually execute. What do you intend to replace it withL

File: cmd/sonobuoy/app/gen.go (hunk 5 above)
Comment (timothysc) [reply cmd/sonobuoy/app/gen.go]: Whoops leftover experiment.

File: cmd/sonobuoy/app/gen.go (hunk 6 above)
Comment (timothysc) [reply cmd/sonobuoy/app/gen.go]: It can't actually or else I would need var at function scope.  

It's used below.

Comment (liztio) [reply cmd/sonobuoy/app/gen.go]: ah, ignore me!

File: cmd/sonobuoy/app/operations/gen.go (hunk 4 above)
Comment (timothysc) [reply cmd/sonobuoy/app/operations/gen.go]: I was planning to expand the number of operations GenConfig struct, such as LoadDefaults.  I could make it more procedural, but every operation would be passing in a GenConfig. 

I was also thinking about the consumer of the library and general UX.  When they include a library, Object[.]F(n) will be auto-discovered vs hunting a function.

File: cmd/sonobuoy/app/gen.go (hunk 9 above)
Comment (timothysc) [reply cmd/sonobuoy/app/gen.go]: I "could" flip the logic, but this reduces more cleanly.

File: cmd/sonobuoy/app/gen.go (hunk 7 above)
Comment (timothysc) [reply cmd/sonobuoy/app/gen.go]: I must be missing something, it is still used by init.

**Comment by timothysc** [comment]
I'll submit follow on PRs for some of the todos and we should do a walk through as well.

**Pull request merged**